gulpfile.js

```diff
@@ -5,21 +5,142 @@
+const omnisharpDownload = require('./out/omnisharpDownload');
+const child_process = require('child_process');
+
+const OmniSharpVersion = omnisharpDownload.OmniSharpVersion;
+
+/// used in offline packaging run so does not clean .vsix
+function clean() {
+    cleanDebugger();
+    return cleanOmnisharp();
+}
+
+gulp.task('clean', ['omnisharp:clean',  'debugger:clean', 'package:clean'], () => {
+
+});
+
+/// Omnisharp Tasks
+function installOmnisharp(omnisharpAssetName) {
+    const logFunction = (message) => { console.log(message); };
+    return omnisharpDownload.downloadOmnisharp(logFunction, omnisharpAssetName);
+}
+
+function cleanOmnisharp() {
+    return del('.omnisharp');
+}
 
 gulp.task('omnisharp:clean', () => {
-	return del('.omnisharp');
+    return cleanOmnisharp();
+});
+ 
+gulp.task('omnisharp:install', ['omnisharp:clean'], () => {
+    var asset = gulpUtil.env.asset || omnisharpDownload.getOmnisharpAssetName();
+    return installOmnisharp(asset);
+});
+
+/// Debugger Tasks
+function getDebugInstaller() {
+    return new debugInstall.DebugInstaller(new debugUtil.CoreClrDebugUtil(path.resolve('.')));
+}
+
+function installDebugger(runtimeId) {
+    return getDebugInstaller().install(runtimeId);
+}
+
+function cleanDebugger() {
+    try {
+        getDebugInstaller().clean();
+        console.log('Cleaned Succesfully');
+    } catch (error) {
+        console.error(error);
+    }
+}
+
+gulp.task('debugger:install', ['debugger:clean'], () => {
+    installDebugger(gulp.env.runtimeId).then(() => {
+        console.log('Installed Succesfully');
+    }).catch((error) => {
+        console.error(error);
+    });
+});
+
+gulp.task('debugger:clean', () => {
+    cleanDebugger();
+});
+
+/// Packaging Tasks
+function doPackage(packageName) {
+
+    var vsceArgs = [];
+    vsceArgs.push(path.join(__dirname, 'node_modules', 'vsce', 'out', 'vsce'))
+    vsceArgs.push('package'); // package command
+
+    if (packageName !== undefined) {
+        vsceArgs.push('-o');
+        vsceArgs.push(packageName);
+    }
+
+    var proc = child_process.spawnSync('node', vsceArgs);
+    if (proc.error) {
+        console.error(proc.error.toString());
+    }
+}
+
+function doOfflinePackage(runtimeId, omnisharpAsset, packageName) {
+    return clean().then(function() {
+        return installDebugger(runtimeId);
+    }).then(function() {
+        return installOmnisharp(omnisharpAsset);
+    }).then(function () {
+        doPackage(packageName + '-' + runtimeId + '.vsix');
+    });
+}
+
+gulp.task('package:clean', () => {
+    return del('*.vsix');
+});
+
+gulp.task('package:online', ['clean'], () => {
+    doPackage();
 });
 
-//TODO: decouple omnisharpDownload (specifically proxy.ts) from vscode 
-// gulp.task('omnisharp:fetch', ['omnisharp:clean'], () => {
-// 	return omnisharpDownload.downloadOmnisharp();
-// });
+gulp.task('package:offline', ['clean'], () => {
+    var json = JSON.parse(fs.readFileSync('package.json'));
```

Comment:
Should we either backup and restore project.json or make sure that the file is clean before and then ask git to revert it?